### PR TITLE
First-pass at a recipe conversion function

### PR DESCRIPTION
**Pull request by schuylermartin45**
This is very much an early POC. There are many `TODO`s and it is far from perfect. There are known bugs and gaps, documented in the `TODO`s

The primary goal of this PR is to implement most of the features discussed in [CEP-13](https://github.com/conda-incubator/ceps/blob/main/cep-13.md) and see how feasible it is to build a conversion utility in percy.

So far, so very good. As noted in the function, I am trying to use as many features of the current parser as possible to:
- Reduce engineering effort
- Gives us flexibility if we decide to expand support past a single conversion function

For less than a day's worth of work, I think this is a pretty solid POC and shows how the parser can make this work less painful.

Annoyingly I can't seem to add @wolfv to the PR despite this being an open project. I will ping him separately, but maybe that's a limitation we need to investigate @jezdez if we want community feedback. I imagine it's tied to some IT permissions group control.

### CODE REVIEW

**Comment by JeanChristopheMorinPerso** [comment]
@schuylermartin45 you can only add reviewers if they are in the "contributor" group if I'm not mistaken.

File: percy/tests/test_aux_files/new_format_simple-recipe.yaml [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+# Comment above a top-level structure
+requirements:
+  empty_field1:
+  host:
+    - if: unix
+      then: setuptools
+    - if: unix
+      then: fakereq
+  empty_field2:  # [unix and win] # selector with comment with comment symbol
```
Comment (schuylermartin45) [inline percy/tests/test_aux_files/new_format_simple-recipe.yaml:22]: This `simple-recipe.yaml` file is used throughout the `percy` unit tests. It is a fake/non-sensical recipe file mean to elicit lots of edge cases. I don't know what meaning a selector on an empty key would really mean/do, but as `CEP-13` doesn't support it, I left this as-is.

File: percy/parser/recipe_parser.py [hunk 2]
```diff
@@ -592,6 +592,93 @@ def render_to_object(self, replace_variables: bool = False) -> JsonType:
...
+
+        # `copy.deepcopy()` produced some bizarre artifacts, namely single-line comments were being incorrectly rendered
+        # as list members. Although inefficient, we have tests that validate round-tripping the parser and there
+        # is no development cost in utilizing tools we already must maintain.
+        new_recipe: RecipeParser = RecipeParser(self.render())
+
+        # Convert the JINJA variable table to a `context` section. Empty tables still add the `context` section for
+        # future developers' convenience.
+        new_recipe.patch(
+            {"op": "add", "path": "/context", "value": new_recipe._vars_tbl if new_recipe._vars_tbl else None}
```
Comment (JeanChristopheMorinPerso) [inline percy/parser/recipe_parser.py:618]: Should we filter out anything that is not a string, number or bool? Because the new format only supports these 3 types in `context`.

https://github.com/conda-incubator/ceps/blob/main/cep-14.md?plain=1#L71

File: percy/tests/test_aux_files/new_format_multi-output.yaml [hunk 3]
```diff
@@ -0,0 +1,29 @@
+context:
+
+outputs:
+  - name: libdb
+    build:
+      run_exports:
```
Comment (wolfv) [inline percy/tests/test_aux_files/new_format_multi-output.yaml:6]: This was moved to `requirements` :)

File: percy/tests/test_aux_files/new_format_simple-recipe.yaml [hunk 4]
```diff
@@ -0,0 +1,56 @@
+context:
```
Comment (JeanChristopheMorinPerso) [inline percy/tests/test_aux_files/new_format_simple-recipe.yaml:1]: The version field is missing at the top of the document.

Comment (JeanChristopheMorinPerso) [reply percy/tests/test_aux_files/new_format_simple-recipe.yaml:1]: It's not mandatory, but why not add it now. We could also add the schema URL as mentioned in https://github.com/conda-incubator/ceps/blob/main/cep-14.md#schema-version.

File: percy/parser/recipe_parser.py (hunk 2 above)
Comment (schuylermartin45) [reply percy/parser/recipe_parser.py:618]: The `RecipeParser` library currently doesn't have logging but this and some of the other edge cases raise the question of if we should add logging. I'm okay with ignoring those types, but I think we need to report it to the user in some capacity.

I suppose I could also return a tuple, one containing a list of errors/warnings. That way I leave the logging up to the calling program.

I also am not 100% confident that parser does properly parse list or dictionary types in JINJA. I don't remember if I have any tests around. There aren't _that_ many recipes that use them.

File: percy/tests/test_aux_files/new_format_multi-output.yaml (hunk 3 above)
Comment (schuylermartin45) [reply percy/tests/test_aux_files/new_format_multi-output.yaml:6]: Isn't that described in one of the later CEPs? This first-pass is focusing on the first CEP. But I'm glad you called it out.

**Comment by schuylermartin45** [comment]
> looks really good! awesome work :)
> 
> If you wire it up with a CLI I'd be excited to try it on a few real-world recipes.

`percy` has some `click` CLIs in the project. I suppose I can bump that ask up in the priority list :)

File: percy/tests/test_aux_files/new_format_simple-recipe.yaml (hunk 4 above)
Comment (schuylermartin45) [reply percy/tests/test_aux_files/new_format_simple-recipe.yaml:1]: Like I said, I focused only on CEP-13, but that sounds like a pretty quick win.

File: percy/parser/recipe_parser.py (hunk 2 above)
Comment (schuylermartin45) [reply percy/parser/recipe_parser.py]: Done.

**Comment by schuylermartin45** [comment]
@wolfv I have added a basic CLI. Assuming you have the `percy` conda environment installed and activated (`make dev` is probably the easiest way to create the environment, the command is:

```sh
percy convert [-o OUTPUT] FILE
```
By default, the recipe is dumped to `STDOUT`. Warnings/Errors always dump to `STDERR`. I don't have any exception handling so if you throw it at a complicated recipe, the parser might blow up on construction.

Here's the output on [a simple recipe](https://github.com/AnacondaRecipes/types-pyyaml-feedstock/blob/main/recipe/meta.yaml)
```yaml
schema_version: 1

context:
  name: types-PyYAML
  version: 6.0.12.12

package:
  name: ${{ name|lower }}
  version: ${{ version }}

source:
  url: https://pypi.io/packages/source/${{ name[0] }}/${{ name }}/types-PyYAML-${{ version }}.tar.gz
  sha256: 334373d392fde0fdf95af5c3f1661885fa10c52167b14593eb856289e1855062

build:
  number: 0
  script: ${{ PYTHON }} -m pip install . --no-deps --no-build-isolation --ignore-installed --no-cache-dir -vv
  skip: ${{ true if py<36 }}

requirements:
  host:
    - pip
    - python
    - wheel
    - setuptools
  run:
    - python

test:
  imports:
    - types
  requires:
    - pip
  commands:
    - pip check
    - if: unix
      then: test -f $SP_DIR/yaml-stubs/__init__.pyi

about:
  home: https://github.com/python/typeshed
  summary: Typing stubs for PyYAML
  license: Apache-2.0 AND MIT
  license_file: LICENSE
  license_family: OTHER
  description: |
    This is a PEP 561 type stub package for the PyYaml package.
    It can be used by type-checking tools like mypy, pyright, pytype,
    PyCharm, etc. to check code that uses PyYaml.
  dev_url: https://github.com/python/typeshed
  doc_url: https://pyyaml.org/wiki/PyYAMLDocumentation

extra:
  recipe-maintainers:
    - fhoehle
    - conda-forge/mypy
```

**Comment by wolfv** [comment]
Awesome! I got it to run with the following `pixi.toml`:

```toml
[project]
name = "percy"
channels = ["conda-forge"]
platforms = ["linux-64", "osx-arm64"]

[tasks]
install = "pip install -e ."

[dependencies]
black = "*"
isort = "*"
make = "*"
mypy = "*"
python = ">=3.11"
pytest = "*"
pytest-cov = "*"
pytest-xdist = "*"
pylint = "*"
pip = "*"
click = ">=8.1.7"
conda = "*"
jinja2 = "*"
pyyaml = "*"
types-pyyaml = "*"
requests = "*"
types-requests = "*"
"ruamel.yaml" = "*"
conda-build = "*"
jsonschema = "*"
types-jsonschema = "*"
pre-commit = "*"
```

You could add more tasks (like lint) and also depenedencies between the tasks can be done (with a `depends_on = ["build"]`...

I'll play around with it! Looks good so far :)

**Comment by schuylermartin45** [comment]
I forgot to return POSIX error codes in the CLI. I aim to get that done today and then I think I'm at the point where I'd like to merge this first iteration in.

**Comment by wolfv** [comment]
Looks great to me! Let me know if you want to go over any additional transformations (of course this can go in a future PR).

For example, the new `tests` section is a bit different (but could be easily adjusted from the previous one, essentially tests need to be split into multiple list elements) :)

**Comment by schuylermartin45** [comment]
@beeankha and I walked through this yesterday offline. Given the general positive conversation in this thread, I'm going to say that this POC is in a good spot to merge.
